### ISSUE TITLE
Open source composition of Edition 3

### ISSUE THREAD

**Issue by kennystrawnmusic** [phil-opp/blog_os#1063]
Progress on the third edition is likely to go much faster if everyone is able to contribute. Although I’ve heard that work on the third edition is indeed in progress, there’s so far nothing visible (yet) — a “work-in-progress” branch of the repository where pre-release contribution to the third edition would be possible would at least get more contributors on board, and the more possible contributors, the faster the development of the third edition of the book would go.

**Comment by phil-opp** [comment]
Thanks for the suggestion and for offering to help!

Right now, the main blocker is that I'm still not satisfied with the new build system. In its current state, we wouldn't be able to use the default `cargo build`/`cargo test`/`cargo run` commands anymore and instead use alias commands (e.g. `cargo kbuild`) to pass the `--target` and `-Zbuild-std` arguments. Another drawback is that we would not be able to use the upcoming [artifact dependencies](https://github.com/rust-lang/rfcs/blob/master/text/3028-cargo-binary-dependencies.md) feature of cargo.

The best solution for this is to support more `.cargo/config` settings in `Cargo.toml` files too, as I [proposed on the internals forum](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336). In order to support `cargo build` etc. with our new build system, we would need to set the following options in our `Cargo.toml`:

- [x] `build.target`
  - Landed in https://github.com/rust-lang/cargo/pull/9030
  - [ ] needs to be fixed: https://github.com/rust-lang/cargo/issues/9451
- [ ] `unstable.build-std`
- [ ] `unstable.build-std-features`
- [ ] `target.runner`

I know that implementing these in cargo isn't an easy task, but it would help very much in bringing the third edition forward.

**Comment by kennystrawnmusic** [comment]
Submitted rust-lang/cargo#10084 as a fix for the showstopper that you mention — rearranges some code around so that the `-Zbuild-std` check waits until **after** `generate_targets()` is already run. Hopefully this solves the problem.

**Comment by phil-opp** [comment]
Thanks a lot for looking into this!

**Comment by kennystrawnmusic** [comment]
@alexchrichton might want to have a one-on-one talk with Phil about this.

**Comment by kennystrawnmusic** [comment]
I would make the point that `target.runner` isn’t *as* critical as `build-std` in Cargo.toml because it can already be set in `.vscode/settings.json` which allows you to just click the “Run” button from VS Code and open it in QEMU that way.

**Comment by kennystrawnmusic** [comment]
Posted a [video](https://youtu.be/lPjoQc822pQ) to my YouTube channel explaining how to get the VS Code option up and running as an alternative just in case the team decides not to move -Zbuild-std into `Cargo.toml` as Alex Crichton’s skepticism might indicate.

**Comment by kennystrawnmusic** [comment]
It also helps to have more creative (and shorter) names for the aliases. I have `cargo bin` set as an alias for building the kernel binary and `cargo boot` set as an alias for building bootable images — much shorter to type than `kbuild`, etc.

**Comment by kennystrawnmusic** [comment]
Alright, since I kind of figured out how to design a logger on my own without the need for Edition 4, I've decided to publish my own logger as a [standalone crate](https://crates.io/crates/printk) for other kernels to use in the meantime. Should help speed up Post 3 by a lot.

**Comment by Lioncat2002** [comment]
yo was wondering how long we got till the publishing of edition 3?

**Comment by kennystrawnmusic** [comment]
> 3?

There is an [edition-3](https://github.com/phil-opp/blog_os/tree/edition-3) repository branch but it hasn’t been updated since January 27 (and I have submitted a few pull requests to that branch myself that have yet to be considered for merging).

**Comment by kennystrawnmusic** [comment]
> yo was wondering how long we got till the publishing of edition 3?

There is an [edition-3](https://github.com/phil-opp/blog_os/tree/edition-3) repository branch but it hasn’t been updated since January 27 (and I have submitted a few pull requests to that branch myself that have yet to be considered for merging).

**Comment by bjorn3** [comment]
@phil-opp is still working on a redesign of the bootloader crate: https://github.com/rust-osdev/bootloader/pull/232 This redesign is blocking the 3rd edition.

**Comment by kennystrawnmusic** [comment]
> @phil-opp is still working on a redesign of the bootloader crate: https://github.com/rust-osdev/bootloader/pull/232 This redesign is blocking the 3rd edition.

So that new version of the bootloader has been released; now what?

**Comment by phil-opp** [comment]
Now that the new bootloader version is ready, I can finally start to write the new posts. I plan to continue my work on the `edition-3` branch and I'll comment here once I have some drafts ready. Then I would appreciate some testing and proofreading!

Note that I'm only working on this project in my spare time, so I can't give you a time estimate.

**Comment by asensio-project** [comment]
Hello @phil-opp, in the first post on the section Disabling the standard library, there is an incoherent text: `` and the --edition 2018 flag specifies that we want to use the [2021 edition](https://doc.rust-lang.org/nightly/edition-guide/rust-2021/index.html) of Rust for our crate. ``.

I think that you should change ``--edition 2018`` to ``--edition 2021``.

**Comment by phil-opp** [comment]
Good catch, thanks! Fixed in 96c8aaf8.

**Comment by kennystrawnmusic** [comment]
Post 2 still uses bootloader 0.10 — so the first thing that needs to happen is updating that post to use 0.11 instead. With an already existing migration guide, shouldn't be too difficult, and I'll definitely be glad to help with any drafts of the updated Post 2 here.

**Comment by phil-opp** [comment]
Yes, this is the next step on my list. I try to get to it this weekend.

**Comment by Lioncat2002** [comment]
Hello! By when can we expect edition 3?

**Comment by kennystrawnmusic** [comment]
Not there yet of course but when Phil gets to the `embedded_graphics` post (last in the series) I have a (warning: untested) [reference implementation](https://github.com/kennystrawnmusic/cryptos/blob/master/src/drm/mod.rs) that might be worth looking at.

**Comment by kennystrawnmusic** [comment]
Looks like all the work just stopped — is anything being done behind the scenes?

**Comment by kennystrawnmusic** [comment]
Another important change worth making would be to [this post](https://github.com/phil-opp/blog_os/blob/edition-3/blog/content/edition-3/posts/02-booting/uefi/index.md): document how to depend on `bootloader-x86_64-common` as a way to speed up custom UEFI loader creation

**Comment by kennystrawnmusic** [comment]
Documentation on how to [load an initrd](https://github.com/rust-osdev/bootloader/commit/3262d16a7471543079920051fa35243d1f9508d1) using the bootloader crate would also be welcome.

**Comment by kennystrawnmusic** [comment]
> Hello! By when can we expect the post 3 for edition 3?
> Since it looks like there is no more a blocker of the bootloader crate👀

Update: it's now possible to [depend on the logger provided by `bootloader-x86_64-common`(https://github.com/kennystrawnmusic/cryptos/commit/54b1c99cea1f31155d9a74715fd6838a5c98fac2) for framebuffer logging — even easier approach than the one I had earlier

**Comment by kennystrawnmusic** [comment]
Going to try using a pull request to fix edition-3-post-2. Should make it much easier.

**Comment by kennystrawnmusic** [comment]
Pull request #1187 submitted. Should hopefully be enough.

**Comment by emanuele-em** [comment]
@phil-opp  I write a writer that use bootloader 0.11 and pixel-based framebuffer if you need help with the blog post

**Comment by kennystrawnmusic** [comment]
Merging rust-osdev/bootloader#360 might be an important step here given that it fixes what could become a potential showstopper if left alone

**Comment by kennystrawnmusic** [comment]
Submitted pull request #1223 to get the ball rolling

**Comment by kennystrawnmusic** [comment]
Finally figured out why #1223 was failing the Zola integration; seems the "more" and "TOC" comments are critical for the build process in that case. Completely green on the CI now, so should be ready to merge.

**Comment by Abelkrijgtalles** [comment]
Question about the embedded_graphics crate in edition 3 post 3: When implementing Display, the rust compiler throws a hell of a lot of errors (see log.txt). What version is used in that post?
Files:
[log.txt](https://github.com/phil-opp/blog_os/files/13920874/log.txt)
[cargotomlkernel.txt](https://github.com/phil-opp/blog_os/files/13920885/cargotomlkernel.txt)
[cargotomlmain.txt](https://github.com/phil-opp/blog_os/files/13920886/cargotomlmain.txt)
[code.txt](https://github.com/phil-opp/blog_os/files/13920887/code.txt)

**Comment by phil-opp** [comment]
@Abelkrijgtalles That post is still work-in-progress and still contains some code examples that I just wrote down without testing. In your case, the issue seems to be that the struct is named `FrameBuffer` instead of `Framebuffer`. You can also check out the [`post-3.3`](https://github.com/phil-opp/blog_os/blob/521d4b2fd243b458771060f2b5a09e363c42cab2/kernel/src/framebuffer.rs) branch, which contains the code for this post.

**Comment by Abelkrijgtalles** [comment]
Thank you! I'm looking out to the release of edition 3!
